### ISSUE TITLE
TensorFlow Hub Module?

### ISSUE THREAD

**Issue by HanGuo97** [google-research/bert#8]
Thanks for releasing BERT!

I'm just wondering if BERT will be available on TensorFlow Hub like ELMO (for either fine-tuning or extracting features)?

**Comment by jacobdevlin-google** [comment]
A TF-hub wrapper around `BertModel` will be available soon, but this module will assume the input  has been tokenized in a compatible with (i.e., with `tokenization.py`). The tricky part is the tokenization.

* The tokenization cannot be implemented with current TF ops
* We can wrap it in a custom op (we also have a C++ implementation available internally), but TF hub has limited support for custom ops.
* It's not trivial to figure out how the API of a tokenization op that will work for everyone. (E.g., how to deal with truncation vs. sliding window for long sequences, how to deal with projection from raw -> tokenized for the labels).

**Comment by loretoparisi** [comment]
@jacobdevlin-google my opinion is that you should go to the C++ implementation, because of its architecture and general purpose application, a BERT standalone version will be the best option definitively (see FastText, StarSpace, etc.). Also as you remarked tokenization it's a separate process, just if we consider double bytes, non word boundary languages, diacritics, etc. this is clearly a separate stage. Of course a TF hub module is welcomed, we are using the ELMO right now as suggested by AllenNLP as well.

**Comment by jacobdevlin-google** [comment]
We don't have any particular recipe for generating sentence embeddings out-of-the-box. Using the output from the "classifier" token is definitely _not_ the right way to go, but in terms of how to combine the per-word feature vectors into a single feature vector probably depends on what you're trying to do. If you're trying to create a feature vector where you don't need to learn any parameters, just averaging all of the word vectors might be the only sensible way to go.

If you can learn new parameters, my goto recipe would be "attention pooling" where the `key` tensor is just a learned weight vector and you attend over all of the BERT feature vectors, and then feed that into one or more fully connected layers.

**Comment by zhaopku** [comment]
@jacobdevlin-google Hi, do you have an estimate of when the TF hub module will be released?

**Comment by jageshmaharjan** [comment]
So, idea is to tokenize the sentence using "tokenization.py" and later run with tf_hub (i.e soon, when available). using the input as a sentence wouldn't work with the tf_hub module, right?.
If we input the token as a list it should follow the standard format as done by "tokenization.py" any other tokenization doesn't work, right?
eg: { [CLS] the man [MASK] to the store [SEP] penguin [MASK] are flight ##less birds [SEP] }

**Comment by loretoparisi** [comment]
@jacobdevlin-google any update for the TF hub module? I'm currently using the Universal Sentence Encoder hub module, but I assume that BERT should give better results in the STS task, etc.
Thank you.

**Comment by miweru** [comment]
I tested BERT with SentEval and a few averaging techniques like tf-idf and weighted averaging and max_pooling.
The out of the box results were not better than fasttext-bow averaged on the STS Tasks.

(The Results in the MR, CR, SUBJ and SST-2 Tasks were better than the best results reported in the SentEval Paper. So the averaging and max_pooling works fine on some tasks.)

On the other hand changing the word-order has strong effects on the resulting averaged vector. (and that produces very exiting results as this simple bow is very sensible to changing the word order)

**Comment by taylorchu** [comment]
@miweru 
Could you share the number and your experiment setup more?

Thanks!

**Comment by hsm207** [comment]
> We don't have any particular recipe for generating sentence embeddings out-of-the-box. Using the output from the "classifier" token is definitely _not_ the right way to go, but in terms of how to combine the per-word feature vectors into a single feature vector probably depends on what you're trying to do. 

@jacobdevlin-google why is using the output from the "classifier" token to create a sentence embedding wrong?

**Comment by miweru** [comment]
> > We don't have any particular recipe for generating sentence embeddings out-of-the-box. Using the output from the "classifier" token is definitely _not_ the right way to go, but in terms of how to combine the per-word feature vectors into a single feature vector probably depends on what you're trying to do.
> 
> @jacobdevlin-google why is using the output from the "classifier" token to create a sentence embedding wrong?

I think it is not trained to be a real representation of the whole sentence, it is useful if we fine tune it on our specific sentence level task. (is it possible to use a typical sentence embedding unsupervised task to train this token to be a good sentence embedding?)

**Comment by hoangcuong2011** [comment]
@miweru : "I think it is not trained to be a real representation of the whole sentence." -> I also think so. I think using the embedding for sentence (via embedding for CLS token) is not enough to learn sentence similarity. Specifically, I think something simple like computing cosine similarity for two sentences based on their two CLS tokens is not good enough.

To bring its best, we need to (1): fine tune it on our specific sentence level task, or simply keep the embedding fixed but have an MLP on top of that so that MLP will be trained for a specific task.

But I was wondering: Is there anyone who has a different opinion on this?

Many thanks!

**Comment by HanGuo97** [comment]
Just realized BERT is now available on the TensorFlow Hub, thanks the team for the effort!

**Comment by AgoloCuongHoang** [comment]
Hi @HanGuo97 Could you help me by writing a short example how to use BERT on the tensorflow hub?  Many thanks!

**Comment by loretoparisi** [comment]
Adding that the example https://github.com/google-research/bert/blob/master/run_classifier_with_tfhub.py is not so clear. 

Why the hub module signature takes encoded input (pre-processed tokens?)

```
bert_inputs = dict(
input_ids=input_ids,
input_mask=input_mask,
segment_ids=segment_ids)
```

**Comment by jageshmaharjan** [comment]
yea, right. I was trying to figure about using, but unable. maybe need some extra task.
+1

**Comment by Armour** [comment]
@AgoloCuongHoang I think TFHub docs has pretty much everything you want: https://tfhub.dev/s?q=bert 

Rememnber that you need to do the pre-processing first before feeding the data, for example, use [WordpieceTokenizer](https://github.com/google-research/bert/blob/master/tokenization.py#L300) to split the input string into a list of wordpiece tokens and map these tokens to `input_ids` using [convert_tokens_to_ids](https://github.com/google-research/bert/blob/master/tokenization.py#L144) function. `input_mask` and `segment_ids` can usually be get easily from the pre-preprocessing pipeline. After that, just follow the docs like below.

```python
# Feature based (used as embedding)
bert_module = hub.Module("https://tfhub.dev/google/bert_uncased_L-12_H-768_A-12/1", trainable=False)
# Fine-tuning BERT model
# bert_module = hub.Module("https://tfhub.dev/google/bert_uncased_L-12_H-768_A-12/1", trainable=True)

bert_inputs = dict(
    input_ids=input_ids,
    input_mask=input_mask,
    segment_ids=segment_ids)
bert_outputs = bert_module(bert_inputs, signature="tokens", as_dict=True)
pooled_output = bert_outputs["pooled_output"]
sequence_output = bert_outputs["sequence_output"]
```

**Comment by loretoparisi** [comment]
@Armour could you please make an example of the pre-processing? Second question, SentencePiece is a better tool than WordPiece, would it be possibile to use [SentencePiece](https://github.com/google/sentencepiece) tokens / ids as input rather than WordPiece?

**Comment by HanGuo97** [comment]
I believe SentencePiece is [not compatible](
https://github.com/google-research/bert#learning-a-new-wordpiece-vocabulary) with WordPiece.

**Comment by hoangcuong2011** [comment]
Hi. If anyone can give us a full example including preprocessing as well, that would be extremely useful! Many thanks!

**Comment by Armour** [comment]
@loretoparisi Hi, I can create a repo for a simple example of using BERT module when I have time, probably on this weekend. 

Yes, you can use SentencePiece if you pre-train your own BERT model with it. Take it in mind that the BERT model from TFHub was pre-trained with WordPieceTokenizer, so I think you should better keep it consistent unless you pre-train your own BERT model from scratch using SentencePiece, the expected result should be similar to the WordPieceTokenizer one. Let me know if I'm wrong :)

For your last question, I think Jacob is a better person to ask and I believe Google will also take that into consideration.

**Comment by jageshmaharjan** [comment]
@hoangcuong2011 , the example is in https://colab.sandbox.google.com/github/google-research/bert/blob/master/predicting_movie_reviews_with_bert_on_tf_hub.ipynb,

**Comment by akshaykgupta** [comment]
Hi, Is there a way to access intermediate layer outputs in the BERT Hub module? I would like to use a weighted sum of average embeddings from each layer (in the style of ELMo).

**Comment by robmsylvester** [comment]
Following akshaykgupta's request, are there any plans to expose these other layers, such as what is done with the inception networks available on tf hub? I've been trying to hack it together a bit but haven't quite stitched it all together. I have asked a question [here] on stack overflow for reference.(https://stackoverflow.com/questions/55333558/how-to-access-bert-intermediate-layer-outputs-in-tf-hub-module)

**Comment by HolmesSherlock** [comment]
Tensorflow Hub's [bert_uncased_L-12_H-768_A-12](https://tfhub.dev/google/bert_uncased_L-12_H-768_A-12/1) documentations says,

> We currently only support the tokens signature,...

However, [Predicting Movie Reviews with BERT on TF Hub](https://colab.research.google.com/github/google-research/bert/blob/master/predicting_movie_reviews_with_bert_on_tf_hub.ipynb) uses the `tokenization_info` signature.

`tokenization_info = bert_module(signature="tokenization_info", as_dict=True)`

Isn't that an inconsistency? How do I know what all are the signatures supported by the TF Hub BERT model?

**Comment by lytum** [comment]
> 
> 
> A TF-hub wrapper around `BertModel` will be available soon, but this module will assume the input has been tokenized in a compatible with (i.e., with `tokenization.py`). The tricky part is the tokenization.
> 
>     * The tokenization cannot be implemented with current TF ops
> 
>     * We can wrap it in a custom op (we also have a C++ implementation available internally), but TF hub has limited support for custom ops.
> 
>     * It's not trivial to figure out how the API of a tokenization op that will work for everyone. (E.g., how to deal with truncation vs. sliding window for long sequences, how to deal with projection from raw -> tokenized for the labeHas the C++ implementation for tokenization been published? Where could I get 

> 
> 
> A TF-hub wrapper around `BertModel` will be available soon, but this module will assume the input has been tokenized in a compatible with (i.e., with `tokenization.py`). The tricky part is the tokenization.
> 
>     * The tokenization cannot be implemented with current TF ops
> 
>     * We can wrap it in a custom op (we also have a C++ implementation available internally), but TF hub has limited support for custom ops.
> 
>     * It's not trivial to figure out how the API of a tokenization op that will work for everyone. (E.g., how to deal with truncation vs. sliding window for long sequences, how to deal with projection from raw -> tokenized for the labels).

where could we get the c++ implemnetation for tokenization? has it been released or not?

**Comment by kyawminar** [comment]
free go to my bold hren - sosi huy!

пн, 14 окт. 2019 г. в 01:45, lytum <notifications@github.com>:
